readium/shared/src/main/java/org/readium/r2/shared/publication/Href.kt

```diff
@@ -0,0 +1,176 @@
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+
+        other as Href
+
+        if (href != other.href) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int =
+        href.hashCode()
+
+    private sealed class Url : Parcelable {
+        abstract fun resolve(base: SharedUrl?, parameters: Map<String, String>): SharedUrl
+        abstract val parameters: List<String>?
+    }
+
+    /**
+     * A static hypertext reference to a publication resource.
+     */
+    @Parcelize
+    private data class StaticUrl(val url: SharedUrl) : Url() {
+        @IgnoredOnParcel
+        override val parameters: List<String>? = null
+
+        override fun resolve(base: SharedUrl?, parameters: Map<String, String>): SharedUrl =
+            base?.resolve(url) ?: url
+
+        override fun toString(): String = url.toString()
+    }
+
+    /**
+     * A templated hypertext reference to a publication resource.
+     *
+     * @param template The URL template, as defined in RFC 6570.
+     */
+    @Parcelize
+    private data class TemplatedUrl(val template: String) : Url() {
+
+        companion object {
+            operator fun invoke(template: String): TemplatedUrl? {
+                // Check that the produced URL is valid.
+                SharedUrl(URITemplate(template).expand(emptyMap())) ?: return null
+                return TemplatedUrl(template)
+            }
+        }
+
+        @IgnoredOnParcel
+        override val parameters: List<String> =
+            URITemplate(template).parameters
+
+        override fun resolve(base: SharedUrl?, parameters: Map<String, String>): SharedUrl {
+            val url = checkNotNull(SharedUrl(URITemplate(template).expand(parameters)))
+            return base?.resolve(url) ?: url
+        }
+
+        override fun toString(): String = template
+    }
+}
+
+/**
+ * Returns a copy of the receiver after normalizing its HREFs to the link with `rel="self"`.
+ */
+@ExperimentalReadiumApi
+public fun Manifest.normalizeHrefsToSelf(): Manifest {
+    val base = linkWithRel("self")?.href?.resolve()
+        ?: return this
+
+    return normalizeHrefsToBase(base)
+}
+
+/**
+ * Returns a copy of the receiver after normalizing its HREFs to the given [baseUrl].
+ */
+@ExperimentalReadiumApi
+public fun Manifest.normalizeHrefsToBase(baseUrl: SharedUrl): Manifest {
+    return copy(HrefNormalizer(baseUrl))
+}
+
+/**
+ * Returns a copy of the receiver after normalizing its HREFs to the given [baseUrl].
+ */
+@ExperimentalReadiumApi
+public fun Link.normalizeHrefsToBase(baseUrl: SharedUrl?): Link {
+    baseUrl ?: return this
+    return copy(HrefNormalizer(baseUrl))
+}
+
+@OptIn(ExperimentalReadiumApi::class)
+private class HrefNormalizer(private val baseUrl: SharedUrl) : ManifestTransformer {
+    override fun transform(href: Href): Href =
+        href.resolveTo(baseUrl)
+}
```

Comment:
I disagree, but I can extract it into its own `HrefNormalizer` file.

Reply:
That would make sense.

In general, I don't like circular dependencies because it requires extra and unnecessary mental bandwidth at reading. I don't want to have to know `Manifest` and `ManifestTransformer` when reading `Link` while I need to know `Manifest` and `ManifestTransformer` reading `Href`.